yarn-project/noir-contracts/src/contracts/inclusion_proofs_contract/src/main.nr

```diff
@@ -0,0 +1,285 @@
+                    },
+                ),
+                public_value: PublicState::new(
+                    context,
+                    2, // Storage slot
+                    FieldSerializationMethods,
+                ),
+            }
+        }
+    }
+
+    #[aztec(private)]
+    fn constructor(public_value: Field) {
+        let selector = compute_selector("_initialize(Field)");
+        context.call_public_function(context.this_address(), selector, [public_value]);
+    }
+
+    #[aztec(public)]
+    internal fn _initialize(value: Field) {
+        storage.public_value.write(value);
+    }
+
+    // Creates a value note owned by `owner`.
+    #[aztec(private)]
+    fn create_note(
+        owner: AztecAddress,
+        value: Field,
+    ) {
+        let owner_private_values = storage.private_values.at(owner.address);
+        let mut note = ValueNote::new(value, owner.address);
+        owner_private_values.insert(&mut note, true);
+    }
+
+
+    // Proves that the owner owned a ValueNote at block `block_number`.
+    #[aztec(private)]
+    fn proveNoteInclusion(
+        owner: AztecAddress,
+        block_number: Field, // The block at which we'll prove that the note exists
+    ) {
+        // TODO: assert that block number is less than the block number of context.block_data
+        // --> This will either require a new oracle method that returns block_data.global_variables_hash preimage
+        //     or modifying the private context so that we somehow expose it.
+
+        // 1) Get historic block data from oracle and ensure that the block hash is included in the current blocks tree
+        //    root.
+        let block_data = get_block_data(block_number, context);
+
+        // 2) Get the note from PXE.
+        let private_values = storage.private_values.at(owner.address);
+        let options = NoteGetterOptions::new().select(1, owner.address).set_limit(1);
+        let notes = private_values.get_notes(options);
+        let note = notes[0].unwrap();
+
+        // 3) Compute the commitment from the note
+        let note_commitment = note_utils::compute_unique_siloed_note_hash(ValueNoteMethods, note);
+
+        // 4) Get the membership witness of the note in the note hash tree
+        let note_hash_tree_id = 2; // TODO(#3443)
+        let witness: MembershipWitness<NOTE_HASH_TREE_HEIGHT, NOTE_HASH_TREE_HEIGHT + 1> =
+            get_membership_witness(block_number, note_hash_tree_id, note_commitment);
+
+        // 5) Prove that the commitment is in the note hash tree
+        assert(
+            block_data.note_hash_tree_root == compute_merkle_root(note_commitment, witness.index, witness.path),
+            "Proving membership of a note in note hash tree failed"
+        );
+
+        // --> Now we have traversed the trees all the way up to blocks tree root.
+    }
+
+    // Proves that the note was not yet nullified at block `block_number`.
+    #[aztec(private)]
+    fn proveNullifierNonInclusion(
```

Comment:
same as above-  can we not move this entire fn to a library?

Reply:
As I mentioned on slack will do this in a separate PR 😊